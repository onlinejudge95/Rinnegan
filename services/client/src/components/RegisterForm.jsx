import React from "react";
import { Redirect } from "react-router-dom";
import { Formik } from "formik";
import * as Yup from "yup";
import { PropTypes } from "prop-types";

const RegisterForm = (props) => {
  if (props.isAuthenticated()) {
    return <Redirect to="/" />;
  }

  return (
    <div>
      <h1 className="title is-1">Register</h1>
      <hr />
      <br />
      <Formik
        initialValues={{
          username: "",
          email: "",
          password: "",
        }}
        onSubmit={(values, { setSubmitting, resetForm }) => {
          props.handleRegisterFormSubmit(values);
          resetForm();
          setSubmitting(false);
        }}
        validationSchema={Yup.object().shape({
          username: Yup.string()
            .required("Username is required")
            .min(6, "Username must be greater than 5 characters"),
          email: Yup.string()
            .required("Email is required")
            .min(6, "Email must be greater than 5 characters"),
          password: Yup.string()
            .required("Password is required")
            .min(11, "Password must be greater than 10 characters"),
        })}
      >
        {(props) => {
          const {
            values,
            touched,
            errors,
            isSubmitting,
            handleBlur,
            handleChange,
            handleSubmit,
          } = props;

          return (
            <form onSubmit={handleSubmit}>
              <div className="field">
                <label className="label" htmlFor="input-username">
                  Username
                </label>
                <input
                  name="username"
                  id="input-username"
                  type="text"
                  placeholder="Enter a Username"
                  value={values.username}
                  onBlur={handleBlur}
                  onChange={handleChange}
                  className={
                    errors.username && touched.username
                      ? "input error"
                      : "input"
                  }
                />
                {errors.username && touched.username && (
                  <div className="input-feedback" data-testid="errors-username">
                    {errors.username}
                  </div>
                )}
              </div>
              <div className="field">
                <label className="label" htmlFor="input-email">
                  Email
                </label>
                <input
                  name="email"
                  id="input-email"
                  type="email"
                  placeholder="Enter an Email address"
                  value={values.email}
                  onBlur={handleBlur}
                  onChange={handleChange}
                  className={
                    errors.email && touched.email ? "input error" : "input"
                  }
                />
                {errors.email && touched.email && (
                  <div className="input-feedback" data-testid="errors-email">
                    {errors.email}
                  </div>
                )}
              </div>
              <div className="field">
                <label className="label" htmlFor="input-password">
                  Password
                </label>
                <input
                  name="password"
                  id="input-password"
                  type="password"
                  placeholder="Enter a Password"
                  value={values.password}
                  onBlur={handleBlur}
                  onChange={handleChange}
                  className={
                    errors.password && touched.password
                      ? "input error"
                      : "input"
                  }
                />
                {errors.password && touched.password && (
                  <div className="input-feedback" data-testid="errors-password">
                    {errors.password}
                  </div>
                )}
              </div>
              <input
                type="submit"
                className="button is-primary"
                value="Submit"
                disabled={isSubmitting}
              />
            </form>
          );
        }}
      </Formik>
    </div>
  );
};

RegisterForm.propTypes = {
  handleRegisterFormSubmit: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.func.isRequired,
};

export default RegisterForm;
