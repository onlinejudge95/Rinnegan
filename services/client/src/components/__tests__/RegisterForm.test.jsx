import React from "react";
import { cleanup, fireEvent, wait } from "@testing-library/react";

import RegisterForm from "../RegisterForm";

afterEach(cleanup);

describe("renders", () => {
  const props = {
    handleRegisterFormSubmit: () => {
      return true;
    },
    isAuthenticated: () => {
      return false;
    },
  };

  it("registration form", () => {
    const dom = renderWithRouter(<RegisterForm {...props} />);
    expect(dom.getByText("Register")).toHaveClass("title");
  });

  it("default props", () => {
    const dom = renderWithRouter(<RegisterForm {...props} />);

    const usernameInput = dom.getByLabelText("Username");
    expect(usernameInput).toHaveAttribute("type", "text");
    expect(usernameInput).not.toHaveValue();

    const emailInput = dom.getByLabelText("Email");
    expect(emailInput).toHaveAttribute("type", "email");
    expect(emailInput).not.toHaveValue();

    const passwordInput = dom.getByLabelText("Password");
    expect(passwordInput).toHaveAttribute("type", "password");
    expect(passwordInput).not.toHaveValue();

    const buttonInput = dom.getByText("Submit");
    expect(buttonInput).toHaveValue("Submit");
  });

  it("a snapshot properly", () => {
    const dom = renderWithRouter(<RegisterForm {...props} />);
    expect(dom.asFragment()).toMatchSnapshot();
  });
});

describe("handles form validation correctly", () => {
  const mockProps = {
    handleRegisterFormSubmit: jest.fn(),
    isAuthenticated: jest.fn(),
  };

  it("when fields are empty", async () => {
    const dom = renderWithRouter(<RegisterForm {...mockProps} />);

    expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);

    const usernameInput = dom.getByLabelText("Username");
    fireEvent.blur(usernameInput);
    expect((await dom.findByTestId("errors-username")).innerHTML).toBe(
      "Username is required"
    );

    const emailInput = dom.getByLabelText("Email");
    fireEvent.blur(emailInput);
    expect((await dom.findByTestId("errors-email")).innerHTML).toBe(
      "Email is required"
    );

    const passwordInput = dom.getByLabelText("Password");
    fireEvent.blur(passwordInput);
    expect((await dom.findByTestId("errors-password")).innerHTML).toBe(
      "Password is required"
    );

    const form = dom.container.querySelector("form");
    fireEvent.submit(form);

    await wait(() => {
      expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);
    });
  });

  it("when email field is not valid", async () => {
    const { getByLabelText, container, findByTestId } = renderWithRouter(
      <RegisterForm {...mockProps} />
    );

    expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);

    const emailInput = getByLabelText("Email");
    fireEvent.change(emailInput, { target: { value: "test_user" } });
    fireEvent.blur(emailInput);
    expect((await findByTestId("errors-email")).innerHTML).toBe(
      "Enter a valid email"
    );

    const form = container.querySelector("form");
    fireEvent.submit(form);

    await wait(() => {
      expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);
    });
  });

  it("when fields are not the proper length", async () => {
    const { getByLabelText, container, findByTestId } = renderWithRouter(
      <RegisterForm {...mockProps} />
    );

    expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);

    const usernameInput = getByLabelText("Username");
    fireEvent.change(usernameInput, { target: { value: "null" } });
    fireEvent.blur(usernameInput);
    expect((await findByTestId("errors-username")).innerHTML).toBe(
      "Username must be greater than 5 characters"
    );

    const emailInput = getByLabelText("Email");
    fireEvent.change(emailInput, { target: { value: "t@t.c" } });
    fireEvent.blur(emailInput);
    expect((await findByTestId("errors-email")).innerHTML).toBe(
      "Email must be greater than 5 characters"
    );

    const passwordInput = getByLabelText("Password");
    fireEvent.change(passwordInput, { target: { value: "invalid" } });
    fireEvent.blur(passwordInput);
    expect((await findByTestId("errors-password")).innerHTML).toBe(
      "Password must be greater than 10 characters"
    );

    const form = container.querySelector("form");
    fireEvent.submit(form);

    await wait(() => {
      expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);
    });
  });

  it("when fields are valid", async () => {
    const { getByLabelText, container, findByTestId } = renderWithRouter(
      <RegisterForm {...mockProps} />
    );

    expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);

    const usernameInput = getByLabelText("Username");
    fireEvent.change(usernameInput, { target: { value: "test_user" } });
    fireEvent.blur(usernameInput);

    const emailInput = getByLabelText("Email");
    fireEvent.change(emailInput, { target: { value: "test_user@mail.com" } });
    fireEvent.blur(emailInput);

    const passwordInput = getByLabelText("Password");
    fireEvent.change(passwordInput, { target: { value: "test_password" } });
    fireEvent.blur(passwordInput);

    const form = container.querySelector("form");
    fireEvent.submit(form);

    await wait(() => {
      expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(1);
    });
  });
});
