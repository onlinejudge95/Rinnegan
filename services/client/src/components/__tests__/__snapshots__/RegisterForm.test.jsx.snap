// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot properly 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="title is-1"
    >
      Register
    </h1>
    <hr />
    <br />
    <form>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-username"
        >
          Username
        </label>
        <input
          class="input"
          id="input-username"
          name="username"
          placeholder="Enter a Username"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-email"
        >
          Email
        </label>
        <input
          class="input"
          id="input-email"
          name="email"
          placeholder="Enter an Email address"
          type="email"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-password"
        >
          Password
        </label>
        <input
          class="input"
          id="input-password"
          name="password"
          placeholder="Enter a Password"
          type="password"
          value=""
        />
      </div>
      <input
        class="button is-primary"
        type="submit"
        value="Submit"
      />
    </form>
  </div>
</DocumentFragment>
`;
